Create driver for PIT and some time functions

xlmod opened the pull request:
PIT (Programmable Interval Timer) generate interrupt at regular interval.
Create a function to set the frequence of this interval and a init function and add it in kernel_init.

Create time relative functions using pit interface. A sleep function that return after at least the number of millisecond given in arguments and a gettime function that return the number of millisecond since kernel the kernel start (in reality it's since the interrupts are enabled).

xlmod committed:
Create driver for PIT and some time functions

PIT (Programmable Interval Timer) generate interrupt at regular
interval.
Create a function to set the frequence of this interval and a init
function and add it in kernel_init.

Create time relative functions using pit interface.
A sleep function that return after at least the number of millisecond
given in arguments and a gettime function that return the number of
millisecond since kernel the kernel start (in reality it's since the
interrupts are enabled).

Signed-off-by: xlmod <glafond-@student.42.fr>

e2r3p13 commented on kernel/time.c:
```diff
@@ -0,0 +1,24 @@
+
+/* kernel/time.c
+ *
+ *
```
file description

e2r3p13 commented on kernel/kernel.c:
```diff
@@ -28,9 +30,15 @@ static void init_descriptor_tables() {
 	idt_init();
 }
 
+static void kernel_init() {
+	pit_init();
+	// TODO: Put all init function here. With pit_init first to get track of time as soon as possible.
+}
+
 void kernel_main(unsigned long multiboot_info_addr) {
 	multiboot_info_t *mbi = (multiboot_info_t *)multiboot_info_addr;
 
+	kernel_init();
 	init_descriptor_tables();
 	VGA_initialize();
 	KBD_initialize();
```
Either move all the init function inside the new `kernel_init` function or move out the `pit_init` in the `kernel_main` and delete the `kernel_init` function.

xlmod replied:
done

e2r3p13 commented on include/kernel/time.h:
```diff
@@ -0,0 +1,27 @@
+
+/* include/kernel/time.h
+ *
+ * Insert file description here
```
Change the file description

e2r3p13 commented on include/kernel/pit.h:
```diff
@@ -0,0 +1,25 @@
+#define PIT_TIMEMS ((pit_counter * 1000) / pit_freq)
+
+extern uint32_t pit_freq;
+extern uint32_t pit_counter;
```
I think this variable should be marked as volatile, because it can be changed asynchronously by an interrupt while polled from a function e.g. `sleep`. Marking it as volatile would avoid the compiler to make assumptions and loop forever --> I'm not 100% sure about it, let's check.

xlmod replied:
done

e2r3p13 commented on drivers/pit/pit.c:
```diff
@@ -0,0 +1,34 @@
+}
+
+void pit_set_freq(uint32_t freq) {
+	uint16_t div = PIT_INTERNAL_FREQ / freq;
```
`div` is a common function name and is likely to be used for a function later e.g. in the math lib. Maybe we should rename it. Not very important now tho.

As a hint of it, some IDEs highlights it as a keyword. take a look at Github's one.

xlmod replied:
fixed

e2r3p13 commented on include/kernel/pit.h:
```diff
@@ -0,0 +1,25 @@
+
+/* include/kernel/pit.h
+ *
+ *
```
file description por favor

e2r3p13 commented on drivers/pit/pit.c:
```diff
@@ -0,0 +1,34 @@
+
+/* drivers/pit/pit.c
+ *
+ *
```
File description s'il vous plait

e2r3p13 reviewed (dismissed):
Some minor adjustments, but the overall looks good

xlmod committed:
Change PIT counter and time calculation functions

Switch from a single counter to two counters:
- a counter representing the interrupt number modulo the frequency.
- a counter representing the interrupt number divided by the frequency.

Add PIT_REALTIMEMS, a macro that calculates the timer drift and returns
the corrected time.

Signed-off-by: glafond- <glafond-@student.42.fr>
